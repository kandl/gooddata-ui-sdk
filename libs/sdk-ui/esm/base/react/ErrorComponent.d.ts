import React from "react";
/**
 * Props of the {@link ErrorComponent}.
 * @public
 */
export interface IErrorProps {
    code?: string;
    icon?: string;
    message: string;
    description?: string;
    className?: string;
    style?: object;
    width?: any;
    /**
     * Size of the error component.
     */
    height?: any;
    /**
     * Size of the visualisation content when custom size layouting is enabled.
     */
    clientHeight?: any;
}
/**
 * Component that renders a default error message.
 *
 * @remarks
 * See {@link https://sdk.gooddata.com/gooddata-ui/docs/error_component.html | ErrorComponent}
 *
 * @public
 */
export declare class ErrorComponent extends React.Component<IErrorProps> {
    static defaultProps: Partial<IErrorProps>;
    render(): JSX.Element;
}
//# sourceMappingURL=ErrorComponent.d.ts.map