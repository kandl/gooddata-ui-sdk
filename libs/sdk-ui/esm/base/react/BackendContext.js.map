{"version":3,"file":"BackendContext.js","sourceRoot":"","sources":["../../../src/base/react/BackendContext.tsx"],"names":[],"mappings":"AAAA,qCAAqC;AACrC,OAAO,KAAK,MAAM,OAAO,CAAC;AAE1B,OAAO,EAAE,eAAe,EAAE,MAAM,sBAAsB,CAAC;AACvD,OAAO,EAAE,SAAS,EAAE,MAAM,cAAc,CAAC;AAEzC,MAAM,cAAc,GAAG,KAAK,CAAC,aAAa,CAAiC,SAAS,CAAC,CAAC;AACtF,cAAc,CAAC,WAAW,GAAG,gBAAgB,CAAC;AAkB9C;;;;GAIG;AACH,MAAM,CAAC,MAAM,eAAe,GAAoC,CAAC,EAAE,QAAQ,EAAE,OAAO,EAAE,EAAE,EAAE;IACtF,OAAO,oBAAC,cAAc,CAAC,QAAQ,IAAC,KAAK,EAAE,OAAO,IAAG,QAAQ,CAA2B,CAAC;AACzF,CAAC,CAAC;AAEF;;;;;;;;;;;;;;;;;;;;GAoBG;AACH,MAAM,CAAC,MAAM,UAAU,GAAG,CAAC,OAA4B,EAAkC,EAAE;IACvF,MAAM,kBAAkB,GAAG,KAAK,CAAC,UAAU,CAAC,cAAc,CAAC,CAAC;IAC5D,OAAO,OAAO,aAAP,OAAO,cAAP,OAAO,GAAI,kBAAkB,CAAC;AACzC,CAAC,CAAC;AAEF;;;;;;;;;;;;;;;;;;;;;;GAsBG;AACH,MAAM,CAAC,MAAM,gBAAgB,GAAG,CAC5B,OAA4B,EAC5B,OAAO,GAAG,kBAAkB,EACV,EAAE;IACpB,MAAM,kBAAkB,GAAG,KAAK,CAAC,UAAU,CAAC,cAAc,CAAC,CAAC;IAC5D,MAAM,gBAAgB,GAAG,OAAO,aAAP,OAAO,cAAP,OAAO,GAAI,kBAAkB,CAAC;IACvD,SAAS,CACL,gBAAgB,EAChB,GAAG,OAAO,oIAAoI,CACjJ,CAAC;IACF,OAAO,gBAAgB,CAAC;AAC5B,CAAC,CAAC;AAEF;;;;;GAKG;AACH,MAAM,UAAU,WAAW,CACvB,SAAiC;IAEjC,MAAM,4BAA4B,GAAgB,CAAC,KAAK,EAAE,EAAE;QACxD,OAAO,CACH,oBAAC,cAAc,CAAC,QAAQ,QACnB,CAAC,OAAO,EAAE,EAAE,WAAC,OAAA,oBAAC,SAAS,oBAAK,KAAK,IAAE,OAAO,EAAE,MAAA,KAAK,CAAC,OAAO,mCAAI,OAAO,IAAI,CAAA,EAAA,CACnD,CAC7B,CAAC;IACN,CAAC,CAAC;IAEF,OAAO,eAAe,CAAC,aAAa,EAAE,SAAS,CAAC,CAAC,4BAA4B,CAAC,CAAC;AACnF,CAAC"}