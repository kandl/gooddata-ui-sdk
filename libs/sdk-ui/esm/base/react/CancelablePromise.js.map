{"version":3,"file":"CancelablePromise.js","sourceRoot":"","sources":["../../../src/base/react/CancelablePromise.ts"],"names":[],"mappings":"AAAA,qCAAqC;AACrC,OAAO,OAAO,MAAM,mBAAmB,CAAC;AAYxC;;GAEG;AACH,MAAM,oBAAoB,GAAG,UAAU,CAAC;AAExC;;GAEG;AACH,MAAM,OAAO,WAAY,SAAQ,KAAK;IAClC,YAAmB,MAAe;QAC9B,KAAK,CAAC,oBAAoB,CAAC,CAAC;QADb,WAAM,GAAN,MAAM,CAAS;QAG9B,gIAAgI;QAChI,MAAM,CAAC,cAAc,CAAC,IAAI,EAAE,WAAW,CAAC,SAAS,CAAC,CAAC;IACvD,CAAC;IAED;;OAEG;IACI,SAAS;QACZ,OAAO,IAAI,CAAC,MAAM,CAAC;IACvB,CAAC;CACJ;AAED;;GAEG;AACH,MAAM,CAAC,MAAM,aAAa,GAAG,CAAC,GAAY,EAAsB,EAAE;IAC9D,OAAO,OAAO,CAAC,GAAG,CAAC,IAAI,GAAG,CAAC,OAAO,KAAK,oBAAoB,CAAC;AAChE,CAAC,CAAC;AAEF;;;;;;;;;;;GAWG;AACH,MAAM,UAAU,cAAc,CAAI,OAAmB;IACjD,IAAI,YAAgC,CAAC;IACrC,IAAI,WAAW,GAAG,KAAK,CAAC;IACxB,IAAI,YAAY,GAAG,KAAK,CAAC;IAEzB,MAAM,cAAc,GAAG,IAAI,OAAO,CAAI,CAAC,OAAO,EAAE,MAAM,EAAE,EAAE;QACtD,OAAO,CAAC,IAAI,CACR,CAAC,KAAK,EAAE,EAAE;YACN,IAAI,WAAW,EAAE;gBACb,MAAM,CAAC,IAAI,WAAW,CAAC,YAAY,CAAC,CAAC,CAAC;aACzC;iBAAM;gBACH,YAAY,GAAG,IAAI,CAAC;gBACpB,OAAO,CAAC,KAAK,CAAC,CAAC;aAClB;QACL,CAAC,EACD,CAAC,KAAK,EAAE,EAAE;YACN,IAAI,WAAW,EAAE;gBACb,MAAM,CAAC,IAAI,WAAW,CAAC,YAAY,CAAC,CAAC,CAAC;aACzC;iBAAM;gBACH,YAAY,GAAG,IAAI,CAAC;gBACpB,MAAM,CAAC,KAAK,CAAC,CAAC;aACjB;QACL,CAAC,CACJ,CAAC;IACN,CAAC,CAAC,CAAC;IAEH,OAAO;QACH,OAAO,EAAE,cAAc;QACvB,MAAM,EAAE,CAAC,MAAe,EAAE,EAAE;YACxB,WAAW,GAAG,IAAI,CAAC;YACnB,YAAY,GAAG,MAAM,CAAC;QAC1B,CAAC;QACD,cAAc,EAAE,GAAG,EAAE,CAAC,WAAW;QACjC,eAAe,EAAE,GAAG,EAAE,CAAC,YAAY;KACtC,CAAC;AACN,CAAC"}