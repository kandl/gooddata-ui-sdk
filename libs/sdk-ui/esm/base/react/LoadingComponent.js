// (C) 2007-2022 GoodData Corporation
import React from "react";
const baseAnimationDuration = 1.4;
/**
 * Component that renders a default loading indicator.
 *
 * @remarks
 * See {@link https://sdk.gooddata.com/gooddata-ui/docs/loading_component.html | LoadingComponent }
 *
 * @public
 */
class LoadingComponent extends React.Component {
    render() {
        const { inline, width, height, imageWidth, imageHeight, color, speed = 1, className } = this.props;
        const duration = baseAnimationDuration / speed;
        const delay = duration / -5;
        const dotStyles = {
            transformOrigin: "4px 4px",
            animation: `GDC-pop ${duration}s infinite`,
            animationDelay: `${delay * 2}s`,
            fill: color,
        };
        const wrapperStyles = {
            textAlign: "center",
            display: inline ? "inline-flex" : "flex",
            verticalAlign: "middle",
            flexDirection: "column",
            alignContent: "center",
            justifyContent: "center",
            flex: "1 0 auto",
            height,
            width,
        };
        const svgStyles = {
            maxHeight: "100%",
            maxWidth: "100%",
            flex: "0 1 auto",
            width: imageWidth,
            height: imageHeight,
        };
        const dot1Styles = dotStyles;
        const dot2Styles = Object.assign(Object.assign({}, dotStyles), { transformOrigin: "18px 4px", animationDelay: `${delay}s` });
        const dot3Styles = Object.assign(Object.assign({}, dotStyles), { transformOrigin: "32px 4px", animationDelay: "0" });
        return (React.createElement("div", { className: className, style: wrapperStyles },
            React.createElement("svg", { style: svgStyles, version: "1.1", x: "0px", y: "0px", viewBox: "0 0 36 8" },
                React.createElement("style", null, `
                        @keyframes GDC-pop {
                            0%,
                            80%,
                            100% {
                                transform: scale(0);
                            }
                            40% {
                                transform: scale(1);
                            }
                        }
                    `),
                React.createElement("g", { style: dot1Styles },
                    React.createElement("circle", { cx: "4", cy: "4", r: "4" })),
                React.createElement("g", { style: dot2Styles },
                    React.createElement("circle", { cx: "18", cy: "4", r: "4" })),
                React.createElement("g", { style: dot3Styles },
                    React.createElement("circle", { cx: "32", cy: "4", r: "4" })))));
    }
}
LoadingComponent.defaultProps = {
    className: "s-loading",
    color: "var(--gd-palette-complementary-6, #94a1ad)",
    speed: 1,
    inline: false,
    height: "100%",
    width: undefined,
    imageHeight: 8,
    imageWidth: undefined,
};
export { LoadingComponent };
//# sourceMappingURL=LoadingComponent.js.map