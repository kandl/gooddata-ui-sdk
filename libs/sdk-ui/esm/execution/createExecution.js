// (C) 2019-2022 GoodData Corporation
import { isAttribute, isMeasure, MeasureGroupIdentifier, newDimension, newTwoDimensional, } from "@gooddata/sdk-model";
import compact from "lodash/compact.js";
import isEmpty from "lodash/isEmpty.js";
import { invariant } from "ts-invariant";
/**
 * When caller desires just data series and no slicing, create a single-dim result.
 */
function seriesOnlyDim(seriesBy) {
    return [
        newDimension(compact([
            ...seriesBy.filter(isAttribute),
            // only include MeasureGroupIdentifier if there are some measures, otherwise the execution will always fail on the backend
            seriesBy.some(isMeasure) && MeasureGroupIdentifier,
        ])),
    ];
}
/**
 * When caller desires data series to be sliced further by some attributes (and perhaps with totals as well)
 * then create two-dim result resembling a pivot table:
 *
 * -  slices are in rows (first dim)
 * -  measures & scoping attributes will be in columns (second dim)
 */
function seriesAndSlicesDim(seriesBy, slices, totals) {
    return newTwoDimensional([...slices, ...totals], compact([
        ...seriesBy.filter(isAttribute),
        // only include MeasureGroupIdentifier if there are some measures, otherwise the execution will always fail on the backend
        seriesBy.some(isMeasure) && MeasureGroupIdentifier,
    ]));
}
/**
 * Given execute props, this will prepare execution to send to backend.
 *
 * @param options - create execution options
 * @internal
 */
export function createExecution(options) {
    const { backend, workspace, seriesBy = [], slicesBy = [], filters = [], sortBy = [], totals = [], componentName = "Execution", } = options;
    invariant(backend && workspace, "backend and workspace must be either specified explicitly or be provided by context");
    const dimensions = isEmpty(slicesBy)
        ? seriesOnlyDim(seriesBy)
        : seriesAndSlicesDim(seriesBy, slicesBy, totals);
    return backend
        .withTelemetry(componentName, options)
        .workspace(workspace)
        .execution()
        .forItems(seriesBy.concat(slicesBy), filters)
        .withSorting(...sortBy)
        .withDimensions(...dimensions);
}
//# sourceMappingURL=createExecution.js.map