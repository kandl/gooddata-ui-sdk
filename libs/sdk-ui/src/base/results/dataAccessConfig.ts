// (C) 2019-2023 GoodData Corporation
import { DataValue } from "@gooddata/sdk-model";
import NumberJs from "@gooddata/numberjs";
import isEmpty from "lodash/isEmpty.js";
import escape from "lodash/escape.js";
import unescape from "lodash/unescape.js";

const customEscape = (str: string) => str && escape(unescape(str));

/**
 * @public
 */
export type ValueFormatter = (value: DataValue, format: string) => string;

/**
 * @public
 */
export type HeaderTranslator = (value: string | null) => string;

/**
 * Creates value formatter that uses `@gooddata/numberjs` to format raw measure values according
 * to the format string.
 *
 * @remarks
 * By default, the format will strip away all the coloring information and
 * just return the value as string.
 *
 * @param separators - number separators to use. if not specified then `numberjs` defaults will be used
 * @public
 */
export function createNumberJsFormatter(separators?: NumberJs.ISeparators): ValueFormatter {
    return (value: DataValue, format: string) => {
        const valueToFormat = value === null && !isEmpty(format) ? "" : parseFloat(value as string);

        const formattedValue = NumberJs.numberFormat(valueToFormat, format, undefined, separators);
        const formattedObject: NumberJs.INumberObject = NumberJs.colors2Object(formattedValue);

        return customEscape(formattedObject.label);
    };
}

/**
 * Default configuration for the data access methods. Uses default `@gooddata/numberjs` formatter and no result formatting.
 *
 * @public
 */
export const DefaultDataAccessConfig: DataAccessConfig = {
    valueFormatter: createNumberJsFormatter(),
};

/**
 * @public
 */
export type DataAccessConfig = {
    /**
     * Function to use to format measure values.
     */
    valueFormatter: ValueFormatter;

    /**
     * Function to translate header names.
     */
    headerTranslator?: HeaderTranslator;
};
