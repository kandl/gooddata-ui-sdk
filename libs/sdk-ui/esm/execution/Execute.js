// (C) 2019-2022 GoodData Corporation
import React from "react";
import { withExecution } from "./withExecution.js";
import isEqual from "lodash/isEqual.js";
import { useResolveValuesWithPlaceholders, withContexts, UnexpectedSdkError, } from "../base/index.js";
import { createExecution } from "./createExecution.js";
const CoreExecute = (props) => {
    const { children, error, isLoading, reload, result, LoadingComponent, ErrorComponent } = props;
    if (ErrorComponent && error) {
        return React.createElement(ErrorComponent, { error: error });
    }
    if (LoadingComponent && isLoading) {
        return React.createElement(LoadingComponent, null);
    }
    if (LoadingComponent && ErrorComponent && !result) {
        return null;
    }
    return children({
        error,
        isLoading,
        reload,
        result,
    });
};
function componentName(props) {
    return props.componentName || "Execute";
}
function exportTitle(props) {
    return props.exportTitle || componentName(props);
}
const WrappedExecute = withContexts(withExecution({
    exportTitle,
    execution: (props) => {
        const { seriesBy, slicesBy, totals, filters, sortBy } = props;
        if (!(seriesBy === null || seriesBy === void 0 ? void 0 : seriesBy.length) && !(slicesBy === null || slicesBy === void 0 ? void 0 : slicesBy.length)) {
            throw new UnexpectedSdkError("In the Execute component, either seriesBy or slicesBy must be defined and must contain at least one item");
        }
        return createExecution(Object.assign(Object.assign({}, props), { componentName: componentName(props), seriesBy: seriesBy, slicesBy: slicesBy, totals: totals, filters: filters, sortBy: sortBy }));
    },
    events: (props) => {
        const { onError, onLoadingChanged, onLoadingFinish, onLoadingStart, onExportReady } = props;
        return {
            onError,
            onLoadingChanged,
            onLoadingFinish,
            onLoadingStart,
            onExportReady,
        };
    },
    shouldRefetch: (prevProps, nextProps) => {
        const relevantProps = [
            "onError",
            "onLoadingChanged",
            "onLoadingFinish",
            "onLoadingStart",
        ];
        const relevantPropsDeepEqual = [
            "seriesBy",
            "slicesBy",
            "totals",
            "filters",
            "sortBy",
            "window",
            "workspace",
        ];
        return (relevantProps.some((propName) => prevProps[propName] !== nextProps[propName]) ||
            relevantPropsDeepEqual.some((propName) => !isEqual(prevProps[propName], nextProps[propName])));
    },
    loadOnMount: (props) => {
        const { loadOnMount = true } = props !== null && props !== void 0 ? props : {};
        return loadOnMount;
    },
    window: (props) => props.window,
})(CoreExecute));
/**
 * The executor provides a more curated experience to obtain and work with data from backends.
 *
 * @remarks
 * It is aligned with the `DataAccess` infrastructure which exposes the underlying data as data series that can be
 * sliced by additional attributes.
 *
 * Once the executor finishes, the {@link DataViewFacade.data} method will expose the data as series and
 * slices according to the specification to the executor.
 * Note that if the resulting data is empty this will NOT throw a NoDataError. It is the responsibility
 * of the child component to handle that if they need to.
 *
 * See {@link IDataAccessMethods} for additional documentation
 * @public
 */
export const Execute = (props) => {
    const [seriesBy, slicesBy, totals, filters, sortBy] = useResolveValuesWithPlaceholders([props.seriesBy, props.slicesBy, props.totals, props.filters, props.sortBy], props.placeholdersResolutionContext);
    return React.createElement(WrappedExecute, Object.assign({}, props, { seriesBy, slicesBy, totals, filters, sortBy }));
};
//# sourceMappingURL=Execute.js.map