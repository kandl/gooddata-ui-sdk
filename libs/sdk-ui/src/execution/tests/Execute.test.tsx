// (C) 2019-2021 GoodData Corporation
import React from "react";
import { mount } from "enzyme";
import { dummyBackend, dummyBackendEmptyData } from "@gooddata/sdk-backend-mockingbird";
import { createDummyPromise } from "../../base/react/tests/toolkit";
import { DataViewFacade } from "../../base/results/facade";
import { IAnalyticalBackend } from "@gooddata/sdk-backend-spi";
import { Execute, IExecuteProps } from "../Execute";
import { ReferenceMd } from "@gooddata/reference-workspace";
import { newAttributeSort, newPositiveAttributeFilter, newTotal } from "@gooddata/sdk-model";
import { createExecution, CreateExecutionOptions } from "../createExecution";
import { LoadingComponent } from "../../base/react/LoadingComponent";
import { IExecuteErrorComponent } from "../interfaces";

const DummyBackendEmptyData = dummyBackendEmptyData();
const makeChild = () => jest.fn((_) => <div />);
const renderDummyExecutor = (
    child: jest.Mock<JSX.Element>,
    props: Omit<IExecuteProps, "backend" | "workspace" | "children" | "seriesBy"> = {},
    backend: IAnalyticalBackend = DummyBackendEmptyData,
) => {
    return mount(
        <Execute backend={backend} workspace={"testWorkspace"} seriesBy={[ReferenceMd.Amount]} {...props}>
            {child}
        </Execute>,
    );
};

describe("Execute", () => {
    it("should start loading immediately and inject isLoading prop", () => {
        const child = makeChild();
        renderDummyExecutor(child);

        expect(child).toHaveBeenCalledWith({
            isLoading: true,
            error: undefined,
            result: undefined,
            reload: expect.any(Function),
        });
    });

    it("should not start loading immediately if loadOnMount is set to false", () => {
        const child = makeChild();
        renderDummyExecutor(child, { loadOnMount: false });

        expect(child).toHaveBeenCalledWith({
            isLoading: false,
            error: undefined,
            result: undefined,
            reload: expect.any(Function),
        });
    });

    it("should stop loading when execution is resolved and inject data view facade", async () => {
        const child = makeChild();
        renderDummyExecutor(child);
        await createDummyPromise({ delay: 100 });

        expect(child).toHaveBeenCalledWith({
            isLoading: false,
            error: undefined,
            result: expect.any(DataViewFacade),
            reload: expect.any(Function),
        });
    });

    it("should stop loading when execution fails", async () => {
        const child = makeChild();
        renderDummyExecutor(child, {}, dummyBackend());
        await createDummyPromise({ delay: 100 });

        expect(child).toHaveBeenCalledWith({
            isLoading: false,
            error: expect.any(Error),
            result: undefined,
            reload: expect.any(Function),
        });
    });

    it("should start loading after invoking injected reload function", async () => {
        const child = jest.fn(({ reload }) => <button onClick={reload} />);
        const wrapper = renderDummyExecutor(child, { loadOnMount: false });
        wrapper.find("button").simulate("click");

        expect(child).toHaveBeenCalledWith({
            isLoading: false,
            error: undefined,
            result: undefined,
            reload: expect.any(Function),
        });
        expect(child).toHaveBeenCalledWith({
            isLoading: true,
            error: undefined,
            result: undefined,
            reload: expect.any(Function),
        });
    });

    it("should invoke onLoadingStart, onLoadingChanged and onLoadingFinish events", async () => {
        const child = makeChild();
        const onLoadingStart = jest.fn();
        const onLoadingChanged = jest.fn();
        const onLoadingFinish = jest.fn();

        renderDummyExecutor(child, {
            onLoadingChanged,
            onLoadingFinish,
            onLoadingStart,
        });

        await createDummyPromise({ delay: 100 });

        expect(onLoadingStart).toBeCalledTimes(1);
        expect(onLoadingChanged).toBeCalledTimes(2);
        expect(onLoadingFinish).toBeCalledTimes(1);
    });

    it("should invoke onError when execution fails with a NoDataError without a DataView", async () => {
        const child = makeChild();
        const onLoadingStart = jest.fn();
        const onLoadingChanged = jest.fn();
        const onLoadingFinish = jest.fn();
        const onError = jest.fn();

        renderDummyExecutor(
            child,
            {
                onError,
                onLoadingChanged,
                onLoadingFinish,
                onLoadingStart,
            },
            dummyBackend({ raiseNoDataExceptions: "without-data-view" }),
        );

        await createDummyPromise({ delay: 100 });

        expect(onError).toBeCalledTimes(1);
        expect(onLoadingStart).toBeCalledTimes(1);
        expect(onLoadingChanged).toBeCalledTimes(2);
        expect(onLoadingFinish).not.toBeCalled();
    });

    it("should NOT invoke onError when execution fails with a NoDataError with a DataView", async () => {
        const child = makeChild();
        const onLoadingStart = jest.fn();
        const onLoadingChanged = jest.fn();
        const onLoadingFinish = jest.fn();
        const onError = jest.fn();

        renderDummyExecutor(
            child,
            {
                onError,
                onLoadingChanged,
                onLoadingFinish,
                onLoadingStart,
            },
            dummyBackend({ raiseNoDataExceptions: "with-data-view" }),
        );

        await createDummyPromise({ delay: 100 });

        expect(onError).not.toBeCalled();
        expect(onLoadingStart).toBeCalledTimes(1);
        expect(onLoadingChanged).toBeCalledTimes(2);
        expect(onLoadingFinish).toBeCalledTimes(1);
    });

    it("should render LoadingComponent", async () => {
        const child = makeChild();
        const dummyExecutor = renderDummyExecutor(child, {
            LoadingComponent,
        });

        expect(dummyExecutor.find(LoadingComponent)).toExist();
    });

    it("should render ErrorComponent, when execution fails", async () => {
        const ErrorComponent: IExecuteErrorComponent = () => <div />;
        const child = makeChild();
        const dummyExecutor = renderDummyExecutor(
            child,
            {
                ErrorComponent,
            },
            dummyBackend(),
        );

        await createDummyPromise({ delay: 100 });

        dummyExecutor.update();

        expect(dummyExecutor.find(ErrorComponent)).toExist();
    });

    it("should not call children function without result, when both Loading & ErrorComponent are provided", async () => {
        const ErrorComponent: IExecuteErrorComponent = () => <div />;
        const child = makeChild();
        renderDummyExecutor(child, {
            LoadingComponent,
            ErrorComponent,
        });

        expect(child).not.toHaveBeenCalledWith(
            expect.objectContaining({
                result: undefined,
            }),
        );

        await createDummyPromise({ delay: 100 });

        expect(child).toHaveBeenCalledWith(
            expect.objectContaining({
                result: expect.any(DataViewFacade),
            }),
        );
    });
});

describe("createExecution", () => {
    const Scenarios: Array<[string, Partial<IExecuteProps>]> = [
        ["unscoped series only", { seriesBy: [ReferenceMd.Amount] }],
        ["scoped series only", { seriesBy: [ReferenceMd.Amount, ReferenceMd.Region] }],
        [
            "unscoped series with slicing",
            { seriesBy: [ReferenceMd.Amount], slicesBy: [ReferenceMd.Product.Name] },
        ],
        [
            "scoped series with slicing",
            { seriesBy: [ReferenceMd.Amount, ReferenceMd.Region], slicesBy: [ReferenceMd.Product.Name] },
        ],
        [
            "scoped series with slicing and filter",
            {
                seriesBy: [ReferenceMd.Amount, ReferenceMd.Region],
                slicesBy: [ReferenceMd.Product.Name],
                filters: [newPositiveAttributeFilter(ReferenceMd.Product.Name, ["CompuSci"])],
            },
        ],
        [
            "scoped series with slicing and sortBy",
            {
                seriesBy: [ReferenceMd.Amount, ReferenceMd.Region],
                slicesBy: [ReferenceMd.Product.Name],
                sortBy: [newAttributeSort(ReferenceMd.Product.Name, "desc")],
            },
        ],
        [
            "scoped series with slicing and totals",
            {
                seriesBy: [ReferenceMd.Amount, ReferenceMd.Region],
                slicesBy: [ReferenceMd.Product.Name],
                totals: [newTotal("sum", ReferenceMd.Amount, ReferenceMd.Product.Name)],
            },
        ],
    ];

    it.each(Scenarios)("should create correct definition for %s", (_desc, props) => {
        const input: IExecuteProps = {
            backend: DummyBackendEmptyData,
            workspace: "testWorkspace",
            seriesBy: props.seriesBy!,
            children: makeChild(),
            ...props,
        };

        const execution = createExecution(input as CreateExecutionOptions);

        expect(execution.definition).toMatchSnapshot();
    });
});
