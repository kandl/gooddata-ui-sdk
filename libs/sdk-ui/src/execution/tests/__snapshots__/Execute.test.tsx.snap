// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createExecution > should create correct definition for scoped series only 1`] = `
{
  "attributes": [
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.owner.region",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.owner.region",
      },
    },
  ],
  "buckets": [],
  "dimensions": [
    {
      "itemIdentifiers": [
        "a_label.owner.region",
        "measureGroup",
      ],
    },
  ],
  "filters": [],
  "measures": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": {},
  "sortBy": [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution > should create correct definition for scoped series with slicing 1`] = `
{
  "attributes": [
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.owner.region",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.owner.region",
      },
    },
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": [],
  "dimensions": [
    {
      "itemIdentifiers": [
        "a_label.product.id.name",
      ],
    },
    {
      "itemIdentifiers": [
        "a_label.owner.region",
        "measureGroup",
      ],
    },
  ],
  "filters": [],
  "measures": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": {},
  "sortBy": [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution > should create correct definition for scoped series with slicing and filter 1`] = `
{
  "attributes": [
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.owner.region",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.owner.region",
      },
    },
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": [],
  "dimensions": [
    {
      "itemIdentifiers": [
        "a_label.product.id.name",
      ],
    },
    {
      "itemIdentifiers": [
        "a_label.owner.region",
        "measureGroup",
      ],
    },
  ],
  "filters": [
    {
      "positiveAttributeFilter": {
        "displayForm": {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "in": {
          "values": [
            "CompuSci",
          ],
        },
      },
    },
  ],
  "measures": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": {},
  "sortBy": [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution > should create correct definition for scoped series with slicing and sortBy 1`] = `
{
  "attributes": [
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.owner.region",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.owner.region",
      },
    },
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": [],
  "dimensions": [
    {
      "itemIdentifiers": [
        "a_label.product.id.name",
      ],
    },
    {
      "itemIdentifiers": [
        "a_label.owner.region",
        "measureGroup",
      ],
    },
  ],
  "filters": [],
  "measures": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": {},
  "sortBy": [
    {
      "attributeSortItem": {
        "attributeIdentifier": "a_label.product.id.name",
        "direction": "desc",
      },
    },
  ],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution > should create correct definition for scoped series with slicing and totals 1`] = `
{
  "attributes": [
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.owner.region",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.owner.region",
      },
    },
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": [],
  "dimensions": [
    {
      "itemIdentifiers": [
        "a_label.product.id.name",
      ],
      "totals": [
        {
          "attributeIdentifier": "a_label.product.id.name",
          "measureIdentifier": "m_aangOxLSeztu",
          "type": "sum",
        },
      ],
    },
    {
      "itemIdentifiers": [
        "a_label.owner.region",
        "measureGroup",
      ],
    },
  ],
  "filters": [],
  "measures": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": {},
  "sortBy": [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution > should create correct definition for unscoped series only 1`] = `
{
  "attributes": [],
  "buckets": [],
  "dimensions": [
    {
      "itemIdentifiers": [
        "measureGroup",
      ],
    },
  ],
  "filters": [],
  "measures": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": {},
  "sortBy": [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution > should create correct definition for unscoped series only with only attributes (RAIL-4265) 1`] = `
{
  "attributes": [
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": [],
  "dimensions": [
    {
      "itemIdentifiers": [
        "a_label.product.id.name",
      ],
    },
  ],
  "filters": [],
  "measures": [],
  "postProcessing": {},
  "sortBy": [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution > should create correct definition for unscoped series with slicing 1`] = `
{
  "attributes": [
    {
      "attribute": {
        "displayForm": {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": [],
  "dimensions": [
    {
      "itemIdentifiers": [
        "a_label.product.id.name",
      ],
    },
    {
      "itemIdentifiers": [
        "measureGroup",
      ],
    },
  ],
  "filters": [],
  "measures": [
    {
      "measure": {
        "definition": {
          "measureDefinition": {
            "item": {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": {},
  "sortBy": [],
  "workspace": "testWorkspace",
}
`;
